S3 cost endpoint

zanchi-r opened the pull request:
This adds an endpoint to retrieve details about the cost and usage of S3

TRAC-660

zanchi-r committed:
s3: retrieve more detailed line item informations

This adds the retrieval of 3 new values when importing the billing data into elasticsearch.
The usageAmout, itemDescription and productName are used by the s3 cost and usage endpoint

zanchi-r committed:
s3_costs: Add an endpoints to retrieve the s3 storage, bandwidth and requests costs

This adds a new endpoint to retrieve the s3 storage, bandwidth ans requests costs along with some usage information for a given time frame.

Solves TRAC-660

zanchi-r committed:
s3: remove human readable fields from elasticsearch

This removes the human readable fields from elasticsearch.
productCode can be used instead of productName and itemDescription is replaced by serviceCode

zanchi-r committed:
s3_costs: replace human readable fields in elasticsearch queries

This replace the usage of human readable fields in elasticsearch queries by their code counterparts

zanchi-r committed:
s3_costs: use QueryArg decorator to handle parameters

This updates the parameter handling and replaces the manual handling by the QueryArg decorators.

zanchi-r committed:
s3_costs: fix the space usage unit

This fixes the space usage unit, replacing GbDay by GbMonth

zanchi-r committed:
s3_costs: add tests

This adds tests for the filters from the elasticsearch request contructor

giubil commented on s3_costs/es_request_constructor.go:
```diff
@@ -0,0 +1,140 @@
+// it crash :
+//	- If the client is nil or malconfigured, it will crash
+//	- If the index is not an index present in the ES, it will crash
+func GetS3SpaceElasticSearchParams(accountList []uint, durationBegin time.Time,
```
The functions `GetS3SpaceElasticSearchParams`, `GetS3RequestsElasticSearchParams` and `GetS3BandwidthElasticSearchParams` have a lot of repeated code into them. Please try merging the repeated code together

giubil commented on s3_costs/s3_costs_route.go:
```diff
@@ -0,0 +1,173 @@
+	)
+	res, err := searchService.Do(ctx)
+	if err != nil {
+		l.Error("Query execution failed : "+err.Error(), nil)
```
This could be the same insance as what was fixed in the PR #37 where if the index does not exists it will return a 500 instead of returning 200 with no data

giubil commented on s3_costs/s3_costs_route.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// makeElasticSearchRequestsRequest prepares and run the request to retrieve requests usage/cost
+func makeElasticSearchRequestsRequest(ctx context.Context, parsedParams esQueryParams,
```
Again, `makeElasticSearchRequestsRequest` and `makeElasticSearchBandwidthRequest` share a lot of the same code. You should merge the redondant code

giubil commented on scripts/add_mapping_es.sh:
```diff
@@ -42,6 +42,13 @@ curl \
 				"resourceId": {
 					"type": "keyword"
 				},
+				"usageAmount": {
```
I don't think the script is used anymore, it was superseded by the adding of the mapping at startup by @schubev, so there is no need to update it

giubil commented on s3_costs/prepare_response.go:
```diff
@@ -0,0 +1,155 @@
+	return buckets[bucketName]
+}
+
+func parseBucketsStorage(buckets bucketsInfo, parsedDocument bucket) {
```
This function does not return anything. Meaning that it will modify arguments passed to it. This is discouraged, as it makes reading the code less clear.
Either change this function to return the changes, or write a comment explaining what is modified

This comment also applies for all instances of functions that does not return modifications

giubil reviewed (changes requested):
Pretty good PR, there are just some minor problems such as redondant code, and function that do not return  its modifications.

Also, your package name `s3_costs` does not really respect Go [naming convention](https://golang.org/doc/effective_go.html#mixed-caps). Underscores are not really used in Go, it should either be `s3Costs` or `s3/costs`. The latter would be more appropriate I think, as we could have other things in the `s3` package

zanchi-r committed:
costs: move s3_costs package to s3/costs

This commit moves the s3_costs package to s3/costs in order to respect the go naming convention and allow for an easier extension of the s3 package

zanchi-r committed:
s3/costs: fixed 500 error when the es index does not exist

This fixes the 500 error when the elasticsearch index does not exist and returns a more comprehensive error

zanchi-r committed:
s3/costs: return the values that are modified

This updates the parsing functions to return the modifications in order to make the code easier to read

zanchi-r committed:
s3/costs: refactor redundant code

This commit removes some redundant code from the s3/costs package

giubil commented on s3/costs/s3_costs_route.go:
```diff
@@ -0,0 +1,178 @@
+		Name:        "accounts",
+		Type:        routes.QueryArgUintSlice{},
+		Description: "The IDs for many AWS account.",
+	}
```
Here you make the `AwsAccountsQueryArg` not optional, but in the function `GetS3UsageAndCostElasticSearchParams` line 63 of file `s3/costs/es_request_constructor.go` you write a condition that looks like handling for optional `AwsAccounts`

```golang
if len(accountList) > 0 {
 		query = query.Filter(createQueryAccountFilter(accountList))
 	}
```

giubil commented on s3/costs/s3_costs_route.go:
```diff
@@ -0,0 +1,178 @@
+	// AwsAccountsQueryArg allows to get the AWS Account IDs in the URL
+	// Parameters with routes.QueryArgs. These AWS Account IDs will be a
+	// slice of Uint stored in the routes.Arguments map with itself for key.
+	AwsAccountsQueryArg = routes.QueryArg{
```
Is it really necessary to have all the queryParams exported ?

giubil commented on s3/costs/s3_costs_route.go:
```diff
@@ -0,0 +1,178 @@
+			routes.QueryArgs{DateBeginQueryArg},
+			routes.QueryArgs{DateEndQueryArg},
+		),
+	}.H().Register("/s3_costs")
```
You modified the package name from `s3_costs` to `s3/costs`. Wouldn't it be better to have the route url reflect that as well ?

giubil reviewed (dismissed):
Just some minor details to fix up, and it should be good to merge

schubev commented on s3/costs/prepare_response.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+// prepareResponse parses the results from elasticsearch and returns a map of buckets with their usage informations
+func prepareResponse(ctx context.Context, resStorage, resRequests, resBandwidthIn, resBandwidthOut *elastic.SearchResult) (interface{}, error) {
```
Don’t repeat yourself! What about something like this?
```go
func prepareResponse(
    ctx context.Context,
    resStorage, resRequests, resBandwidthIn, resBandwidthOut *elastic.SearchResult
) (interface{}, error) {
    var buckets = make(bucketsInfo)
    var err error
    var components = [...]struct{k string, sr *elastic.SearchResult}{
        { "storage", resStorage },
        { "requests", resRequests },
        { "bandwidthIn", resBandwidthIn },
        { "bandwidthOut", resBandwidthOut },
    }
    for _, cpn := range components {
        buckets, err = parseESResult(ctx, buckets, cpn.sr, cpn.k)
        if err != nil {
           return nil, err
        }
    }
    return buckets, nil
}
```

zanchi-r replied:
Done!

zanchi-r committed:
s3/costs: make the accounts parameter optional

This makes the accounts url parameter optional

zanchi-r committed:
s3/costs: move the /s3/costs

Move the costs endpoint from /s3_costs to /s3/costs

zanchi-r committed:
s3/costs: do not export the url query params

schubev commented on s3/costs/s3_costs_route.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+// getS3CostData returns the s3 cost data based on the query params, in JSON format.
+func getS3CostData(request *http.Request, a routes.Arguments) (int, interface{}) {
```
Same here, please don’t repeat yourself.

zanchi-r replied:
Done!

schubev commented on s3/costs/s3_costs_route.go:
```diff
@@ -0,0 +1,178 @@
+	// Parameters with routes.QueryArgs. These AWS Account IDs will be a
+	// slice of Uint stored in the routes.Arguments map with itself for key.
+	AwsAccountsQueryArg = routes.QueryArg{
+		Name:        "accounts",
```
Elsewhere in the project, we used `aas` as the name for this parameter, please change to this so as to have a consistent API.

zanchi-r replied:
Done

zanchi-r committed:
s3/costs: reduce duplicated code

This reduces duplicated code in the getS3CostData and prepareResponse functions by iterating over the components

zanchi-r committed:
s3/costs: updated account query arg name to be consistent with other routes
